MapboxCoreNavigation/MGLMapboxEvents.swift

```diff
@@ -0,0 +1,165 @@
+        modifiedEventDictionary["durationRemaining"] = round(routeProgress.durationRemaining)
+        
+        modifiedEventDictionary["rerouteCount"] = session.numberOfReroutes
+
+        modifiedEventDictionary["volumeLevel"] = Int(AVAudioSession.sharedInstance().outputVolume * 100)
+        modifiedEventDictionary["screenBrightness"] = Int(UIScreen.main.brightness * 100)
+
+        modifiedEventDictionary["batteryPluggedIn"] = UIDevice.current.batteryState == .charging || UIDevice.current.batteryState == .full
+        modifiedEventDictionary["batteryLevel"] = UIDevice.current.batteryLevel >= 0 ? Int(UIDevice.current.batteryLevel * 100) : -1
+        modifiedEventDictionary["applicationState"] = UIApplication.shared.applicationState.telemString
+        
+        //modifiedEventDictionary["connectivity"] = ??
+        
+        return modifiedEventDictionary
+    }
+}
+
+extension UIApplicationState {
+    var telemString: String {
+        get {
+            switch self {
+            case .active:
+                return "Foreground"
+            case .inactive:
+                return "Inactive"
+            case .background:
+                return "Background"
+            }
+            return "Unknown"
+        }
+    }
+}
+
+extension UIDevice {
+    var machine: String {
+        get {
+            var systemInfo = utsname()
+            uname(&systemInfo)
+            let machineMirror = Mirror(reflecting: systemInfo.machine)
+            let identifier = machineMirror.children.reduce("") { identifier, element in
+                guard let value = element.value as? Int8, value != 0 else { return identifier }
+                return identifier + String(UnicodeScalar(UInt8(value)))
+            }
+            
+            return identifier
+        }
+    }
+}
+
+extension CLLocation {
+    var dictionary: [String: Any] {
+        get {
+            var locationDictionary:[String: Any] = [:]
+            locationDictionary["lat"] = coordinate.latitude
+            locationDictionary["lng"] = coordinate.latitude
+            locationDictionary["altitude"] = altitude
+            locationDictionary["timestamp"] = timestamp.ISO8601
+            locationDictionary["horizontalAccuracy"] = horizontalAccuracy
+            locationDictionary["verticalAccuracy"] = verticalAccuracy
+            locationDictionary["course"] = course
+            locationDictionary["speed"] = speed
+            return locationDictionary
+        }
+    }
+}
+
+class FixedLengthBuffer<T> {
+    private var objects = Array<T>()
+    private var length: Int
+    
+    public init(length: Int) {
+        self.length = length
+    }
+    
+    public func push(_ obj: T) {
+        objects.append(obj)
+        if objects.count == length {
+            objects.remove(at: 0)
+        }
+    }
+    
+    public var allObjects: Array<T> {
+        get {
+            return Array(objects)
+        }
+    }
+}
+
+class CoreFeedbackEvent: Hashable {
```

Comment:
If neither this class nor any of its subclasses need to be public (to Objective-C code), we should make them structs. Swift will synthesize initializers for these structs, they’ll take up [less memory](https://developer.apple.com/videos/play/wwdc2016/416/), and they’re just easier to work with than value classes.